### PR TITLE
Feature: Do not mine while node is initially syncing Stacks blocks

### PR DESCRIPTION
**Pull request by kantai**
This is one method for addressing #2364 -- during initial bootup, a `stacks-node` instance will query a set of "height hint" peers. The node will _not_ begin mining until it reaches the max height returned by those peers.

This PR also sets a default set of height hint nodes, as well as providing more proactive default bootstrap node selection for mainnet configurations (i.e., if the `bootstrap_node` field is _not_ set in the config, but the config is a mainnet config, use the default set).

### CODE REVIEW

**Comment by diwakergupta** [comment]
> This is one method for addressing #2364 -- during initial bootup, a `stacks-node` instance will query a set of "height hint" peers. The node will _not_ begin mining until it reaches the max height returned by those peers.
> 
> This PR also sets a default set of height hint nodes, as well as providing more proactive default bootstrap node selection for mainnet configurations (i.e., if the `bootstrap_node` field is _not_ set in the config, but the config is a mainnet config, use the default set).

I'm curious if we expect scenarios where one would want *distinct* sets of "height hint" nodes that are different from bootstrap nodes. Put another way, what's the downside of always using bootstrap nodes provide height hints?

**Comment by kantai** [comment]
> I'm curious if we expect scenarios where one would want distinct sets of "height hint" nodes that are different from bootstrap nodes. Put another way, what's the downside of always using bootstrap nodes provide height hints?

There's two reasons to separate them out (both of which could probably be solved differently):

1. The `rpc_port` and `p2p_port` are configurable separately. So, conceivably, while the bootstrap nodes listen on `p2p_port = 20444`, they may use a port like `20445` for RPC rather than the default `20443`.
2.  Making this configurable allows a miner to turn off the feature by setting the height_hint_nodes to `""`

**Comment by lgalabru** [comment]
I'm also curious about this angle, does that mean that we can't safely rely on the boolean `ibd`, returned by `pox_watchdog.pox_sync_wait`?

**Comment by kantai** [comment]
> I'm also curious about this angle, does that mean that we can't safely rely on the boolean ibd, returned by pox_watchdog.pox_sync_wait?

Yes -- when testing on my mock miner, the node usually tries to start mining once it reaches block height ~120 or so, and takes about another 20 minutes or so to reach the chain tip. So if I were really mining, that'd be 2-3 commits on very old chain tips.

File: testnet/stacks-node/src/config.rs [hunk 1]
```diff
@@ -292,8 +292,16 @@ impl ConfigFile {
             "02afeae522aab5f8c99a00ddf75fbcb4a641e052dd48836408d9cf437344b63516@seed-1.mainnet.stacks.co:20444",
             "03652212ea76be0ed4cd83a25c06e57819993029a7b9999f7d63c36340b34a4e62@seed-2.mainnet.stacks.co:20444"].join(",");
 
+        let height_hint_nodes = [
+            "seed-0.mainnet.stacks.co:20443",
+            "seed-1.mainnet.stacks.co:20443",
+            "seed-2.mainnet.stacks.co:20443",
+        ]
+        .join(",");
```
Comment (jcnelson) [inline testnet/stacks-node/src/config.rs:300]: So, I'm not convinced that we need `height_hint_nodes` at all.  This information is already available to you -- the `PeerNetwork` struct already tracks each of its neighbors' highest known burnchain block heights in its `ConversationP2P` structs.  You ought to be able to add a method to `PeerNetwork` that just goes through the set of `ConversationP2P` structs and extracts the highest reported burn block height (if there are any nodes connected at all).

**Review by jcnelson — CHANGES_REQUESTED** [review]
I don't think this approach is the right way to solve this problem.  I think it is would be simpler and less error-prone to just pull the network's highest known burnchain block height from the p2p thread, since it's already tracking this information for you.

File: testnet/stacks-node/src/config.rs (hunk 1 above)
Comment (kantai) [reply testnet/stacks-node/src/config.rs:300]: But we are trying to get _Stacks_ height, not burnchain block heights. Using the PeerNetwork struct would require instantiating a new network object (and driving the events in it) during startup. That seems way more error prone than just directly querying the `v2/info` endpoint.

**Comment by kantai** [comment]
> I don't think this approach is the right way to solve this problem. I think it is would be simpler and less error-prone to just pull the network's highest known burnchain block height from the p2p thread, since it's already tracking this information for you.

We're trying to obtain the Stacks height, not the burnchain height. Using the P2P thread would allow _any_ node in the network to increase this highest known Stacks height, which would be a bad outcome-- there is a fair amount of trust involved here: if a node reported a very high height, it would cause the miners to not mine.

**Comment by jcnelson** [comment]
> Yes -- when testing on my mock miner, the node usually tries to start mining once it reaches block height ~120 or so, and takes about another 20 minutes or so to reach the chain tip. So if the node were really mining, that'd be 2-3 commits on very old chain tips.

So, the `PoxSyncWatchdog` returns `false` if it detects that (1) it doesn't need to download any more burnchain blocks, and (2) it detects that the node hasn't downloaded or processed any new blocks for a time.  I noticed this phenomenon as well, and I think it's due to the fact that the Stacks node currently just stops processing Stacks blocks altogether until the next burnchain block arrives (I fixed this in the WAL PR; @kantai commented on it).  Can you apply that small fix here and see if driving Stacks block processing makes this problem go away by itself?

**Comment by jcnelson** [comment]
The consensus from the architecture meeting was to make the target Stacks block height a command-line option.  Then, users who want to pull the height from a node can do so with the usual shell command substitution, or they can supply their own.

**Comment by kantai** [comment]
Updated to use `--mine-at-height` command line argument instead.

**Review by jcnelson — APPROVED** [review]
LGTM!  Thanks @kantai

**Review by lgalabru — APPROVED** [review]
Looks good to me!

**Comment by MaksimalistT** [comment]
Thank you for working on this issue, this is really important
just a few comments:
>Updated to use --mine-at-height command line argument instead.
in this case miners would have to guess at what height miner will complete sync
hit lower - wasted bids
hit higher - wasted profits

could you please consider implementing commands to the process, like bitcoin-cli does
user will active mining when it consider miner is ready just send command for ex.: 
>stacks-node-cli start_mine
and for lazy ones we can add argument from which height to mine
>stacks-node-cli start_mine --at-height=1600

it would be great to do the same to preserve already loaded data in working_dir
>stacks-node-cli stop
process gracefully finish i/o operations and leave data in working dir not corrupted, so miner can start not from the scratch

**Pull request merged**